cachi2/core/package_managers/pip.py

```diff
@@ -1411,12 +1423,29 @@ def _download_dependencies(
 
         if require_hashes or req.kind == "url":
             hashes = req.hashes or [req.qualifiers["cachito_hash"]]
-            _verify_hash(download_info["path"], hashes)
+            _verify_hash(download_info["path"], list(map(_to_checksum_info, hashes)))
 
         download_info["kind"] = req.kind
-        downloads.append(download_info)
+        downloaded.append(download_info)
+
+    if allow_binary:
+        log.info("Downloading %d wheel(s) ...", len(to_download))
+        files: dict[str, Union[str, PathLike[str]]] = {pkg.url: pkg.path for pkg in to_download}
+        asyncio.run(async_download_files(files, get_config().concurrency_limit))
+
+        for pkg in to_download:
+            try:
+                key = (canonicalize_name(pkg.name), canonicalize_version(pkg.version))
+                user_req_checksums = user_checksums.get(key, set())
+                checksums = pkg.checksums if pkg.checksums else user_req_checksums
+                _verify_hash(pkg.path, checksums)
```

Comment:
It would be good to tell the user which checksums we're verifiying, e.g.

```python
if pkg.checksums and user_req_checksums:
    checksums = pkg.checksums.intersection(user_req_checksums)
    log.debug("Verifying that <filename> matches both the user-specified hashes and the hashes reported by PyPI")
elif pkg.checksums:
    checksums = pkg.checksums
    log.debug("Verifying that <filename> matches the hashes reported by PyPI")
elif user_req_hashes:
    checksums = pkg.checksums
    log.debug("Verifying that <filename> matches the user-specified hashes")
else:
    # don't verify the hash, _verify_hash(..., []) is always a failure
```

Reply:
I noticed that `_verify_hash` already logs a message about verifying the checksum, maybe not to duplicate the messages (e.g. by not using `_verify_hash`, which is now just a trivial wrapper around `must_match_any_checksum`)

Reply:
[https://github.com/containerbuildsystem/cachi2/pull/255#discussion_r1297132645](related to)

Reply:
[https://github.com/containerbuildsystem/cachi2/pull/255#discussion_r1297436729](url)

Reply:
[https://github.com/containerbuildsystem/cachi2/pull/255#discussion_r1297436729](url)